### ISSUE TITLE
Mysterious detached-row exceptions

### ISSUE THREAD

**Issue by kristiandupont** [realm/realm-dotnet#746]
A number of users have been experiencing detached row exceptions in data binding scenarios. This started appearing after the release of 0.77.0/0.77.1.

**Comment by UKDeveloper99** [comment]
I'm going to make another attempt to reproduce it in my test project when I've a got a spare moment.

**Comment by kristiandupont** [comment]
@UKDeveloper99 and @CrimsonKyle, is there any way either of you could share your code with us? It will be treated completely confidentially and we can sign an NDA if necessary. Alternatively, could we set up a screen sharing session where one of us could debug on your machine? We have not been able to reproduce the bug so far.

**Comment by CrimsonKyle** [comment]
@kristiandupont I just got the green light for a screen share web conference with you to allow us to debug the issue. What times work best for your team? We are in central standard time.

**Comment by UKDeveloper99** [comment]
@kristiandupont I've reproduced it in a test project, uploading it to my dropbox now.

**Comment by UKDeveloper99** [comment]
@kristiandupont  The more test projects the better 😄

**Comment by AndyDentFree** [comment]
Great news, thanks. If you can't post the dropbox URLs here please send them to help@realm.io and Kristian and I can look at them.

**Comment by AndyDentFree** [comment]
I have spent today doing a _partitioning exercise_ separating out the IList changes and DetachedRow detection in 0.77.1 from all the other changes, so with these test projects we should be able to get to a quick answer.

**Comment by AndyDentFree** [comment]
Thanks @UKDeveloper99 I have triggered the error on my machine with your sample.

**Comment by CrimsonKyle** [comment]
We are sending our test project link now @UKDeveloper99 can you send us the link to yours so we can ensure it is replicating the issue in the same way?

**Comment by UKDeveloper99** [comment]
@CrimsonKyle What's your e-mail and I'll send you the link.

**Comment by CrimsonKyle** [comment]
Kyle.Richard93@gmail.com ill also share our dropbox test project with you

**Comment by UKDeveloper99** [comment]
Got it, you can delete that post now if you don't want your e-mail hanging around on here.

**Comment by AndyDentFree** [comment]
@CrimsonKyle have reproduced with clicking on different items in the list (in IOS sim)

**Comment by AndyDentFree** [comment]
This is interesting/frustrating - the two samples are utterly different and the problem comes up with radically different stack crawls.

@UKDeveloper99 I get a crash getting a string property at the bottom of a lot of data binding through MVVMCross for pure IOS.

@CrimsonKyle I get a crash trying to load related list of dogs, in a Xamarin Forms UI.

**Comment by AndyDentFree** [comment]
In both cases, the problem is triggered by having RealmObjects around after the Realm from which they were obtained has gone out of scope. In the Forms UI, it is triggered when you go down into a nested list then come back up, with `GetInstance` called again resetting the field in that GUI object.

In @UKDeveloper99 's `Sandbox`, the Realm is only going into a local variable but objects are passed out of that context.

There is also a problem happening in @CrimsonKyle's sample which may be an unrelated problem, but with the same cause - `Realm` instances are getting cleaned up by the Fnalizer **running on a different thread**.

**Comment by UKDeveloper99** [comment]
@AndyDentFree Is there an easy fix? I'm guessing the same thing was happening but Realm wasn't picking it up in 0.76.1?

**Comment by AndyDentFree** [comment]
in the sample from @CrimsonKyle use the Realm in the App class

1. make `_realm` in the `App` class `internal` access
2. In `RootPage.cs` and `DetailList.cs` use that instance:

```
_realm = ((realmIssue746.App)App.Current)._realm;  
// instead of _realm = Realm.GetInstance();
```
 
In the Sandbox sample from @UKDeveloper99 it was a little harder to work out where to put things but the same style of fix works - just hang onto the Realm instance for the duration.

This is not necessarily the best architecture but to prove robustness, I added a singleton in `KittenModel.cs`

```

    public static class KittenDB
    {
        private static Realm _realm;
        public static Realm db {
            get {
                if (_realm == null)
                    _realm = Realm.GetInstance("KittensRealm.realm");
                return _realm;
            }
        }
    }

```

Then getting an instance used that property:

```
var db = KittenDB.db;  
// instead of var db = Realm.GetInstance("KittensRealm.realm");
```

**Comment by nirinchev** [comment]
@AndyDentFree while this will probably work as a crude workaround, isn't it a little dangerous to expose realms as statics, considering they may be accessed from any thread? I'm just concerned about people grabbing that snippet and then hitting concurrency issues.

Since there are a few people to hit that problem, perhaps the docs should indicate it more explicitly that Realm instances should stay alive for at least as long as the objects retrieved from them.

**Comment by AndyDentFree** [comment]
@nirinchev **Yes** you make a great point - I just wanted to get these folks over an immediate hump whilst we are digging deeper into the _why_ of this (these?) bugs. We need more samples and documentation on best practices. 

We didn't have any integration testing which covered the scenarios here, of a Realm instance being disposed whilst objects from it are still live and are still trying to work out exactly how that is happening.

**Comment by Shazbot89** [comment]
@AndyDentFree Is the work around to **"...use the Realm in the App class"** to detach the instance from the lifecycle of Xam.Forms pages? Trying to make sure I understand.

**Comment by AndyDentFree** [comment]
The workaround is to keep the Realm around for as long as you have objects referring to it.

A cleaner way is to have a ViewModel class that owns a Realm instance.

We are working very hard on this because it _should just work_ and a much smaller subset of the 0.77.1 changes has been identified as triggering this behaviour.

**Comment by CrimsonKyle** [comment]
@AndyDentFree creating the below class and adding the following to the constructor has solved 90% of our issues for the last 10% we are still breaking with the "detached row error message" when our background update process runs and updates the realm while we are accessing a RealmResult that has been converted to a list of realm objects in a UI element but that might be more related to how our sync gateway works and not realm.

```
public static class RealmDB
	{
		private static Realm _realm;
		public static Realm db
			{
				get
				{
					if (_realm == null)
						_realm = Realm.GetInstance();
					return _realm;
				}
			}
	}
```

Then every class we access realm we put the following in the constructor:

```
try
{
	_realm = RealmDB.db;
}
catch (Exception e)
{
	Console.WriteLine(e);
}
```

**Comment by CrimsonKyle** [comment]
Just wanted to say with 0.77.2 it looks like it fixed everything much thanks! I know you all put a lot of work into getting ILists working so great job!

**Comment by Harvinator** [comment]
Just wanted to add my experience for anyone coming here with a rogue "Attempted to access detached row" error.

My problem was that my ViewController had an event listener to a static object which wasn't being removed. (I'd put the code to remove the handler in ViewDidUnload (which doesn't always get fired).

So what was happening was the RealmObject in question was being removed from the realm after the view was closed. Then if i opened the ViewController again the event was actually being fired on both the new & old instance of the ViewController (and the old one was referencing the removed RealmObject)

So a good thing that realm complained about it or I wouldn't have noticed the memory leak (although it was a difficult to debug).

**Comment by aaronjrenfroe** [comment]
I am getting this out of nowhere. I have been using Realm In a project for about a month now and just started getting this exception today. It occurs when my app resumes from the background. I saw that I was on Realm 1.4.0 and Just updated to 1.5.0 and the error is still persisting. I have my realm objects in an observable collection. Before today I was not keeping my realm instance around, after this error I tried to keep it as Global in my class and the error persists.
